### ISSUE TITLE
Floating windows become buried

### ISSUE THREAD

**Issue by alethiophile** [qtile/qtile#3120]
Currently, it seems that when a window is floated, it is removed from the current group layout and (initially) floated on top. Thus, layout commands will only switch between the windows remaining in the layout, and the only way to select a floating window is via the mouse. However, floating windows can also become buried under tiled windows if those windows are raised above them (e.g. if newly created). In this situation, it seems there is no way to bring the floating window to the top again.

Floating windows should probably be unconditionally displayed above tiled windows, regardless of keyboard focus.

**Comment by m-col** [comment]
This is certainly a behaviour that could be improved within qtile. However I don't think the solution is to keep floating windows on top.

You may find something like this useful:
```py
from libqtile import lazy

@lazy.function
def float_to_front(qtile) -> None:
    """Bring all floating windows of the group to front"""
    for window in qtile.current_group.windows:
        if window.floating:
            window.cmd_bring_to_front()
```
If you bind `float_to_front` to a key you can use it to bring all floating windows in the current group above the tiling windows.

If you would like keybindings that can move between all visible windows, check out this module from the examples repo, which may be of interest: https://github.com/qtile/qtile-examples/blob/master/traverse.py

**Comment by alethiophile** [comment]
I have in fact seen that code snippet. It is probably a useful workaround.

However, it seems that this "floating window buried behind tiled window" state is basically just a broken state. If something can easily fall into a broken state, the correct answer is not "here, with this obscure customization you can bind a key to manually un-break it". The answer is to stop it from being broken in the first place.

If there is actually some use for the "floating windows behind tiled windows" state, then the default configuration should be set up to handle it gracefully without requiring any customization, and documentation should be updated to explain it. As it stands, in a default configuration, this is just a land mine that anyone who's trying to use floating windows for any use case other than popups will inevitably run into and have no answer for.

**Comment by jwijenbergh** [comment]
> However, it seems that this "floating window buried behind tiled window" state is basically just a broken state. If something can easily fall into a broken state, the correct answer is not "here, with this obscure customization you can bind a key to manually un-break it". The answer is to stop it from being broken in the first place.

I wouldn't go as far as to say it's a broken state. And I wouldn't even say it's obscure customization. Qtile is a sort of "build it yourself" wm similar to xmonad and awesome, if you don't like some behaviour you can probably customize it.

> If there is actually some use for the "floating windows behind tiled windows" state, then the default configuration should be set up to handle it gracefully without requiring any customization, and documentation should be updated to explain it. As it stands, in a default configuration, this is just a land mine that anyone who's trying to use floating windows for any use case other than popups will inevitably run into and have no answer for.

We could add it the snipper from @m-col to the FAQ.

**Comment by m-col** [comment]
I agree - people's setups and workflows vary a *lot* and so for some people (such as myself) having floating windows at the back can be desirable. I use the above snippet the get out of that state when I need to.

If keeping floating windows above tiled windows is what you're after than a hook might be more appropriate:
```py
from libqtile import hook, qtile

@hook.subscribe.focus_change
def _():
    for window in qtile.current_group.windows:
        if window.floating:
            window.cmd_bring_to_front()
````
This is also a bit hacky and is not perfect. The 'proper' approach would be to mark floating windows to be kept above others - this is part of an ongoing PR for proper Z-axis handling: https://github.com/qtile/qtile/pull/2194

We can certainly do a better job at documenting things like this. Adding a snippet like the top one to the FAQ is an option. Any feedback about what/how things could be documented better is much appreciated.

**Comment by alethiophile** [comment]
As it stands, currently, in a default config it's downright easy to accidentally get floating windows behind tiled windows -- it happens with ordinary focus changes or if any new tiled window is created. As far as I can tell, in a default config there is no way to bring the floating windows back to the top -- the only way I found to address this case is to send all tiled windows away to a different group, then close the floating window (since sending the tiled windows back to the original group would just leave them on top again).

Thus, the default config is definitively broken. If that hook would fix the problem, then it needs to be in the default config, not just in the FAQ. It is a little hacky, and marking floating windows as KEEP_ABOVE by default would presumably be better (once that PR is in).

I remain somewhat confused as to how exactly qtile envisions the default way to deal with floating windows. Is it purely via the mouse, as the default config would seem to suggest? In that case, floating windows really need to be kept on top, because if they go below then they're out of reach of mouse manipulation. Are there other commands that could be bound to keys, that allow switching back to floating windows? If so, they should be bound as part of the default config and mentioned in the documentation.

I know and appreciate that qtile is primarily a "build-your-own" WM that allows universal customization -- that's why I'm here as well. But it's still a problem if the default is broken.

**Comment by cnekmp** [comment]
I see that I'm not the one who has same issues with floating windows. There are some applications that get in "blocking" state when the floating window misclicked and you cannot do anything until you move away tiling windows and get that floating window. What about some config option like `floats_always_ontop = true`?

**Comment by m-col** [comment]
> What about some config option like floats_always_ontop = true?

I agree that this would be ideal, however it is impossible until we have proper Z axis handling (which is a work in progress).

> Thus, the default config is definitively broken. If that hook would fix the problem, then it needs to be in the default config, not just in the FAQ. It is a little hacky, and marking floating windows as KEEP_ABOVE by default would presumably be better (once that PR is in).
> 
> I remain somewhat confused as to how exactly qtile envisions the default way to deal with floating windows. Is it purely via the mouse, as the default config would seem to suggest? In that case, floating windows really need to be kept on top, because if they go below then they're out of reach of mouse manipulation. Are there other commands that could be bound to keys, that allow switching back to floating windows? If so, they should be bound as part of the default config and mentioned in the documentation.

All great points - thanks for the feedback. I agree 100% with what you're saying. I think a sensible default would be to keep floating windows on top by default, using something like a `floats_always_ontop` variable. In the meantime, until the Z axis PR is merged, do you think it be best to include a keybinding to the `float_to_front` defined above in the default config, or perhaps that *and* the hook, one commented out, describing the difference in behaviour?

**Comment by davidbudzynski** [comment]
I've come across the same problem, and I couldn't get my head around why floating windows **are broken** in qtile. Having a keybinding in the default config would definitely be a good mitigation while we're waiting for the PR to be merged.

**Comment by elParaguayo** [comment]
Should we have a command that cycles through every window in the group, focusing them in turn and, if it's a floating window, bring it to the front?

**Comment by davidbudzynski** [comment]
I think that would be the best way to go about this. See also this relevant
issue: https://github.com/qtile/qtile/issues/3120

**Comment by davidbudzynski** [comment]
Sorry, didn't pay attention you mentioned the issue I linked in the topic!

**Comment by davidbudzynski** [comment]
I think having these commands would be beneficial, I can imagine many people get tripped by this, especially those coming from different WMs. I can't imagine why would anyone want to have floating windows behind tiled ones. Likewise, I think this should be the other way round, as it currently is: you should be able to cycle through all windows in the group, the floating windows should stay on top and if you don't like this, you can write a config which  puts floating windows behind tiled ones.

**Comment by strangr** [comment]
please add cmd_keep_above. this is much needed functionality

**Comment by elParaguayo** [comment]
> please add cmd_keep_above. this is much needed functionality

It's in the pull request that mcol referenced.

**Comment by 3nprob** [comment]
Not stale. Recent progress in #3409

**Comment by davidbudzynski** [comment]
Still ongoing issue

**Comment by ramnes** [comment]
This is fixed in 0.23.0 right?
